lib/versions.nix

```diff
@@ -35,15 +40,69 @@ rec {
   */
   patch = v: builtins.elemAt (splitVersion v) 2;
 
+  /* Get string of the first n parts of a version string.
+
+    Example:
+    - truncate 2 "1.2.3-stuff"
+      => "1.2"
+    - truncate 4 "1.2.3-stuff"
+      => "1.2.3.stuff"
+  */
+
+  inherit truncate;
+
   /* Get string of the first two parts (major and minor)
      of a version string.
 
      Example:
        majorMinor "1.2.3"
        => "1.2"
   */
-  majorMinor = v:
-    builtins.concatStringsSep "."
-    (lib.take 2 (splitVersion v));
+  majorMinor = truncate 2;
+
+  /* Get string of the first three parts (major, minor and patch)
+     of a version string.
+
+    Example:
+      majorMinorPatch "1.2.3-stuff"
+      => "1.2.3"
+  */
+  majorMinorPatch = truncate 3;
+
+  /* Version comparison predicates,
+    - isGe v0 v <-> v is greater or equal than v0   [*]
+    - isLe v0 v <-> v is lesser  or equal than v0   [*]
+    - isGt v0 v <-> v is strictly greater than v0   [*]
+    - isLt v0 v <-> v is strictly lesser  than v0   [*]
+    - isEq v0 v <-> v is equal to v0                [*]
+    - range low high v <-> v is between low and high [**]
+
+  [*]  truncating v to the same number of digits as v0
+  [**] truncating v to low for the lower bound and high for the upper bound
+
+    Examples:
+    - isGe "8.10" "8.10.1"
+      => true
+    - isLe "8.10" "8.10.1"
+      => true
+    - isGt "8.10" "8.10.1"
+      => false
+    - isGt "8.10.0" "8.10.1"
+      => true
+    - isEq "8.10" "8.10.1"
+      => true
+    - range "8.10" "8.11" "8.11.1"
+      => true
+    - range "8.10" "8.11+" "8.11.0"
+      => false
+    - range "8.10" "8.11+" "8.11+beta1"
+      => false
+  */
+  isGe = opTruncate versionAtLeast;
+  isGt = opTruncate (flip versionOlder);
+  isLe = opTruncate (flip versionAtLeast);
+  isLt = opTruncate versionOlder;
+  isEq = opTruncate preds.equal;
+  range = low: high: preds.inter (versions.isGe low) (versions.isLe high);
```

Comment:
I find the naming here pretty confusing and it is non obvious to a reader in another file that the version number gets truncated before the predicate is applied.

Also what is the exact motivaton for this `versionAtLeast` etc. can [already deal with versions of different lengths](https://nixos.org/manual/nix/stable/#ssec-version-comparisons).

Reply:
The major difference is that `versionAtLeast` implements an order relation on version numbers (assuming the only separator which is used is `.`), while the predicates I implement here are not relations, they are not "symmetric" (in the sense that  both `isGe "1.1" "1.1.2"` and `isLe "1.1.2" "1.1"` are true) and `(isLe x) y` should be read as `y` is a version which is lower or equal to `x`. Contrarily to `versionAtLeast`, its analogous `isLe` uses the first arguments to determine the length to truncate so that `isLe "1.1" "1.1.2"` is true while `versionAtLeast "1.1" "1.1.2"` is false...

The main motivation is to be able to combine and read them nicely with high order function such as
- `filter`, e.g. the result of `filter (versionAtLeast "1.0") [ "1.3.1" "1.0.0" "1.0" "0.9.0" "1.0.2" ]`, which is `[ "1.0" "0.9.0" ]` is very unnatural to me, both in reading `versionAtLeast` the wrong way around and the fact that I would like `1.0.2` to be included in the output: indeed "1.0.2" is part of the family of "1.0" versions! In comparison `filter (isLe "1.0") [ "1.3.1" "1.0.0" "1.0" "0.9.0" "1.0.2" ]` reads more naturally (we keep versions that are lesser or equal to "1.0") and the outcome `[ "1.0.0" "1.0" "0.9.0" "1.0.2" ]` matches my expectations: if I wanted to filter more precisely I would give the patch version of the first argument of `isLe` instead, and if I want to exclude all "1.0", then I use `isLt "1.0"`.
- `switch` (defined in this PR), for example https://github.com/NixOS/nixpkgs/blob/3daa06cc8b9cc54dd71c004c70921043dcb37105/pkgs/development/coq-modules/mathcomp/default.nix#L21-L22

Reply:
I agree this should be explained in a better way inside the comment in the code. I will try my best to be clear.

Reply:
I pushed a new commit with better explanations (basically what is up here modulo transformation from github PR comment to inline code comment).